### PR TITLE
auth: move PSK handling to samples

### PR DESCRIPTION
**Pull request by sam-golioth**
- move system settings into samples

- add PSK support to hardcoded_credentials module

- remove PSK handling in system_client

- update samples to use the same auth for consistency

### CODE REVIEW

File: samples/settings/src/main.c
Comment (mniestroj) [inline samples/settings/src/main.c:72]: How about calling `hardcoded_credentials_set()` automatically during boot by `SYS_INIT(..., APPLICATION, ...)`? That way we would not need to have that as part of every sample and would keep sample specific code minimal.

File: samples/common/Kconfig
Comment (mniestroj) [inline samples/common/Kconfig:108]: `default n` is implicit, so no need to set that (this line can be removed). Also it is quite hard to which logical change this is related, since all changes are in single commit.

File: samples/hello/README.rst
Comment (mniestroj) [inline samples/hello/README.rst:34]: Please do this kind of improvement in separate commit, so it is less overwhelming to review logical changes.

File: samples/hello/README.rst
Comment (mniestroj) [inline samples/hello/README.rst:231]: No strong opinion here... but I'm a bit afraid of circular referencing (docs are written based this SDK, while here we start to reference those docs), since this page is strictly Golioth Zephyr SDK (this repo) related. It might quickly get outdated.

File: samples/hello/prj.conf
Comment (mniestroj) [inline samples/hello/prj.conf:10]: Is this a leftover to be removed?

File: samples/settings/src/main.c
Comment (sam-golioth) [reply samples/settings/src/main.c:72]: I like it!

File: samples/hello/README.rst
Comment (sam-golioth) [reply samples/hello/README.rst:231]: Good call out.

File: samples/hello/prj.conf
Comment (sam-golioth) [reply samples/hello/prj.conf:10]: Yes 🤦

File: samples/settings/prj.conf
Comment (szczys) [reply samples/settings/prj.conf:16]: > enable storing PSK-ID/PSK in settings subsystem, so that the DFU sample can be flashed on multiple devices as is and then do a DFU of the fleet

Yes, this is how precompiled binaries for the zephyr-training, thingy91demo, and all the reference designs currently work. We're using PSK auth but it's not hardcoded into the firmware.

File: net/golioth/Kconfig
Comment (szczys) [reply net/golioth/Kconfig:314]: Most notable in my mind is persistent storage for WiFI credentials. Currently we use the shell to set to those in the storage partition.

File: samples/settings/prj.conf
Comment (sam-golioth) [reply samples/settings/prj.conf:16]: Hardcoded PSKs are just the default. As part of this PR, there's a new overlay config file in `samples/common` that can enable runtime PSKs via the settings subsystem for _any_ of our samples, e.g.:
`west build -p -b esp32 samples/dfu -- -DOVERLAY_CONFIG="../common/runtime_psk.conf"`

So this should give us the best of both worlds, while maintaining consistency among our samples. All the samples use hardcoded PSKs by default, and all can be changed to use runtime PSKs and the settings shell with a single build argument. This is documented in the READMEs for each of the samples now, e.g.: https://github.com/golioth/golioth-zephyr-sdk/pull/406/files#diff-40c1e6da1ab07e4c146dae9c4f8ecf4c2dd4afdf703d11f21d10dc6b2e4898c3R46

File: net/golioth/Kconfig
Comment (mniestroj) [reply net/golioth/Kconfig:314]: Upstream Zephyr has now support for connection/network manager with some functionalities implemented like management of WiFi (or other medium type) interface. Such "generic" functionalities make most sense to be included as part of Zephyr, since those are not cloud-provider (Golioth) specific. We should make use of those as much as possible, so that:
a) we don't maintain too much utilities that are not directly related to Golioth as a platform
b) it makes less friction for users that want to integrate their existing (or under development) app into Golioth

In other words, it would be great to provide Golioth SDK users some utilities, even for downstream applications. But in those cases it makes more sense to upstream such feature and let it be improved by larger community. Development of such feature (if it is well-thought) should take the same time in both cases. So in most cases some example fits best in `samples/common/` IMO.

In case of PSK-ID/PSK TLS credentials, I think that the ideal situation would be to provide a "TLS credentials storage which utilizes Zephyr settings subsystem". The only thing that would be Golioth-specific is the TLS ID which is used to reference those credentials. This TLS ID is already configured in Kconfig. The rest of it make sense to be put in Zephyr upstream. We can create something temporarily here in SDK until a "proper" implementation is merged to Zephyr, so that we don't introduce any latency between development and being able to use it.

Settings shell is one example why above strategy with upstreaming would be beneficial. First a settings shell was created in Golioth SDK. Shortly after that a different implementation (but with similar functionality in mind) was merged to Zephyr. Then there was a conflict between Kconfig symbols. It was solved (by prepending "our" service with `GOLIOTH_` in Kconfig option name). Right now there is still a conflict in shell module, so either one is visible or the other, since both commands are named `settings`.

Comment (sam-golioth) [reply net/golioth/Kconfig:314]: I like this a lot! I agree, upstreaming is the way to go. 

I filed https://github.com/golioth/firmware-issue-tracker/issues/183 and https://github.com/golioth/firmware-issue-tracker/issues/184 for some of the work that you mentioned, feel free to edit/add/delete.

> We can create something temporarily here in SDK until a "proper" implementation is merged to Zephyr

Where should that live? Still `samples/common` or do we make another folder for things that are in the process of being upstreamed?

**Review by mniestroj — APPROVED** [review]
LGTM, in terms of code changes.

Would be good to have @golioth/developer-relations approval (doesn't have to be here).

**Comment by sam-golioth** [comment]
> LGTM, in terms of code changes.
> 
> Would be good to have @golioth/developer-relations approval (doesn't have to be here).

Confirmed with @ChrisGammell

**Pull request merged**
